ISOLATE_STDLIB_MODULES.rst

```diff
@@ -0,0 +1,241 @@
+   may do the job. For larger modules with lots of types, and possibly also
+   external library callbacks, multiple PR's will be needed.
+
+
+Part 2: Implementation
+......................
+
+Note: this is a suggested implementation plan for a complex module, based on
+lessons learned with other modules. Feel free to simplify it for
+smaller modules.
+
+1. Add Argument Clinic where possible; it enables you to easily use the
+   defining class to fetch module state from type methods.
+2. Prepare for module state; establish a module state `struct`, add an instance
+   as a static global variable, and create helper stubs for fetching the module
+   state.
+3. Add relevant global variables to the module state `struct`, and modify code
+   that accesses the global state to use the module state helpers instead. This
+   step may be broken into several PR's.
+4. Where necessary, convert heap types to static types.
+5. Convert the global module state struct to true module state.
+6. Implement multi-phase initialisation.
+
+Steps 4 through 6 should preferably land in a single alpha development phase.
+
+
+Backwards Compatibility
+=======================
+
+Extension modules in the standard library will now be loadable more than once.
+For example, deleting such a module from ``sys.modules`` and re-importing it
+will result in a fresh module instance, isolated from any previously loaded
+instances.
+
+This may affect code that expected the previous behavior: globals of
+extension modules were shallowly copied from the first loaded module.
+
+
+Security Implications
+=====================
+
+None known.
+
+
+How to Teach This
+=================
+
+A large part of this proposal is a HOWTO aimed at experienced users,
+which will be moved to the documentation.
+
+Beginners should not be affected.
+
+
+Reference Implementation
+========================
+
+Most of the changes are now in the main branch, as commits for these issues:
+
+- `bpo-40077, Convert static types to heap types: use PyType_FromSpec() <https://bugs.python.org/issue40077>`_
+- `bpo-40077, Clear static types in Py_Finalize() for embedded Python <https://bugs.python.org/issue46417>`_
+- `bpo-1635741, Py_Finalize() doesn't clear all Python objects at exit <https://bugs.python.org/issue1635741>`_
+
+As an example, changes and fix-ups done in the `_csv` module are:
+
+- `GH-23224, Remove static state from the _csv module <https://github.com/python/cpython/pull/23224>`_
+- `GH-26008, Allow subclassing of ``csv.Error`` <https://github.com/python/cpython/pull/26008>`_
+- `GH-26074, Add GC support to _csv heap types <https://github.com/python/cpython/pull/26074>`_
+- `GH-26351, Make heap types converted during 3.10 alpha immutable <https://github.com/python/cpython/pull/26351>`_
+
+
+Rejected Ideas
+==============
+
+Do not isolate modules in the standard library
+----------------------------------------------
+
+XXX Someone write something here please
+
+
+Open Issues
+===========
+
+See XXX above.
```

Comment:
This is already covered in the heap types section, right? I guess we could remove this then.